### DISCUSSION TITLE
How do I add cargo features?

### DISCUSSION THREAD

**Discussion by thomas-k-cameron** [smithy-lang/smithy-rs#2182] [Q&A] [answered]
How do I add new dependencies and features?

I understand that `Cargo.toml` is generated with `class CargoTomlGenerator`, which is called by `ClientCodegenVisitor` from `RustClientCodegenPlugin`.

But couldn't really nail down where all the configurations are coming from.

draft pull request: 
https://github.com/awslabs/smithy-rs/pull/2183

UPDATE:

I'm trying to add this piece on the generated toml.
```toml
[target.'cfg(aws_sdk_unstable)'.dependencies]
serde = { version = "1.0", features = ["derive"] }
```

I figured out that you had to add `DependencyScope` and `CargoDependency` so I decided to add these on [CodeDependency.kt](https://github.com/awslabs/smithy-rs/pull/2183/files#diff-e3b8acdd259c52cf69c2229cc7ed43015c0bc1cb141b7a9bf6df69d37758a141) file.
```kotlin
// sealed class DependencyScope
object CfgUnstable : DependencyScope()
```
```kotlin
// data class CargoDependency
val Serde = CargoDependency("serde", CratesIo("1.0"), features = setOf("derive"), scope = DependencyScope.CfgUnstable)
```

and this on [CargoTomlGenerator.kt](https://github.com/awslabs/smithy-rs/pull/2183/files#diff-e3b8acdd259c52cf69c2229cc7ed43015c0bc1cb141b7a9bf6df69d37758a141) file,
```kotlin
"target.'cfg(aws_sdk_unstable)'.dependencies" to dependencies.filter { it.scope == DependencyScope.CfgUnstable }
    .associate { it.name to it.toMap() },
```

But I can't figure out how to tell kotlin to include Serde defined on CargoDependency on CargoTomlGenerator.

**Comment by crisidev** [comment]
If you are asking how a new crate, let's say `hyper` gets written in the `Cargo.toml` of a generated SDK, the answer is that it is done lazily.

When the `RustWriter` interpolates the `CargoDependency`, it will add the new dependency to the available `dependencies` and will fill them in the generated `Cargo.toml`.

**Reply by thomas-k-cameron** [reply]
Sorry, I think my question was not well-formulated.

I have updated the question; I really appreciate if you could check it out!

Thank you in advance.

**Comment by Velfi** [comment, marked as answer]
>But I can't figure out how to tell kotlin to include Serde defined on CargoDependency on CargoTomlGenerator.
It should just automatically add `Serde` once you reference it in a writer:

```kotlin
writer.rust("use #T::{Serialize, Deserialize};", CargoDependency.Serde.toType())
```
